tests/chainermn_tests/functions_tests/test_point_to_point_communication.py

```diff
@@ -12,221 +11,251 @@
-
-    @chainer.testing.attr.gpu
-    def test_tuple_communication1_gpu(self):
-        self.setup(True)
-        self.check_tuple_communication(1)
-
-    @chainer.testing.attr.gpu
-    def test_tuple_communication2_gpu(self):
-        self.setup(True)
-        self.check_tuple_communication(2)
-
-
-class TestNonVariableInput(unittest.TestCase):
-
-    def setUp(self):
-        self.communicator = chainermn.create_communicator('naive')
-
-        if self.communicator.size < 2:
-            pytest.skip('This test is for multinode')
-
-        self.rank_send = (self.communicator.rank + 1) % self.communicator.size
-        self.rank_recv = (self.communicator.rank - 1) % self.communicator.size
-
-    def test_non_variable_send(self):
-        """Checks if backward will be called even if inputs are not Variable.
-
-        This test confirms whether deadlock occurs when numpy/cupy array is
-        given as an input of send.
-        In this case, the input will be converted to chainer Variable without
-        ``requires_grad``, thus ``backward`` will not be called without any
-        modification.
-        """
-        if self.communicator.rank == 0:
-            x = numpy.ones((1, 10)).astype(numpy.float32)
-            phi = chainermn.functions.send(
-                x, self.communicator, rank=self.rank_send)
-            x, = chainermn.functions.pseudo_connect(phi, x)
-            y = chainer.functions.sum(x)
-            t = numpy.array(0).astype(numpy.float32)
-            z = chainer.functions.mean_squared_error(y, t)
-            z.backward()
-
-        elif self.communicator.rank == self.communicator.size - 1:
-            x = chainermn.functions.recv(
-                self.communicator, rank=self.rank_recv)
-            y = chainer.functions.sum(x)
-            t = numpy.array(0).astype(numpy.float32)
-            z = chainer.functions.mean_squared_error(y, t)
-            z.backward()
-
-        else:
-            x = chainermn.functions.recv(
-                self.communicator, rank=self.rank_recv)
-            phi = chainermn.functions.send(
-                x, self.communicator, rank=self.rank_send)
-            phi.backward()
+
+params = [Param(p) for p in [
+    {
+        'dtype': numpy.float16,
+    }, {
+        'dtype': numpy.float32,
+    }]]
+
+
+def create_communicator(gpu, param):
+    if gpu:
+        communicator = chainermn.create_communicator('hierarchical')
+        device = communicator.intra_rank
+        chainer.cuda.get_device_from_id(device).use()
+    else:
+        communicator = chainermn.create_communicator('naive')
+        device = -1
+
+    if communicator.size < 2:
+        pytest.skip('This test is for multinode')
+
+    return communicator
+
+
+def check_communication(gpu, param):
+    communicator = create_communicator(gpu, param)
+    rank_send = (communicator.rank + 1) % communicator.size
+    rank_recv = (communicator.rank - 1) % communicator.size
+    train = TrainEnv(gpu, param, communicator)
```

Comment:
I don't think `TrainEnv` is intuitive. I think we concluded `create_model`-ish name was more intuitive.
`f` and `evaluation` are not a variable so they don't necessarily have to be held in the object.

Reply:
Say, like 
```python
from chainer.functions import sigmoid
import chainer.functions.mean_squared_error as mse
```